catwalk/models/gpt.py

```diff
@@ -1,34 +1,164 @@
-from catwalk.task import Task
 from catwalk.model import Model
+from catwalk.task import InstanceFormat, Task
+from catwalk.tasks.huggingface import HFQAInstance
+
+from torch import log_softmax
+from torch.nn.utils.rnn import pad_sequence
+from transformers import AutoModelForCausalLM, AutoTokenizer
 
 
 @Model.register("catwalk::gpt")
 class GPTModel(Model):
     def __init__(self, pretrained_model_name_or_path: str):
         self.pretrained_model_name_or_path = pretrained_model_name_or_path
+        
+    def _convert_instances(self, instances: Sequence[Dict[str, Any]], instance_format, task) -> MappedSequence:
+        return MappedSequence(lambda instance: task.convert_instance(instance, instance_format), instances)
 
     def predict(  # type: ignore
+        self,
+        task: Task,
+        instances: Sequence[Dict[str, Any]],
+        *,
+        batch_size: int = 32,
+    ) -> Iterator[Dict[str, Any]]:
+        if task.has_instance_conversion(InstanceFormat.HF_QA):
+            return self._predict_qa(task, instances, batch_size=batch_size)
+
+        return self._predict_preplixity(task, instances, batch_size=batch_size)
+
+    def _predict_qa(
+            self,
+            task: Task,
+            instances: Sequence[Dict[str, Any]],
+            *,
+            batch_size: int = 32
+    ) -> Iterator[Dict[str, Any]]:
+        device = resolve_device()
+        model = cached_transformers.get(
+            AutoModelForCausalLM, self.pretrained_model_name_or_path, False).eval().to(device)
+        tokenizer = cached_transformers.get_tokenizer(
+            AutoTokenizer, self.pretrained_model_name_or_path)
+        tokenizer.pad_token = tokenizer.eos_token
+        tokenizer.padding_side = 'left'
+        
+        converted_instances = self._convert_instances(instances, InstanceFormat.HF_QA, task)
+
+        def format_instance(instance: HFQAInstance) -> Tuple[str, str]:
+            # TODO: Use promptsource to add more prompt options?
+            return instance.context, f"\nQuestion:{instance.question}\nAnswer:"
+
+        converted_instances = Tqdm.tqdm(converted_instances, desc="Processing instances")
+        for batch in more_itertools.chunked(converted_instances, batch_size):
+            formatted_batch = [format_instance(instance) for instance in batch]
+            encodings = tokenizer(formatted_batch,
+                                  padding="longest",
+                                  truncation="only_first",
+                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],  # TODO: Make this configurable?
+                                  stride=128,
+                                  return_overflowing_tokens=True,
+                                  return_offsets_mapping=True,
+                                  return_token_type_ids=True,
+                                  return_tensors="pt"
+                                  )
```

Comment:
With the way windowing works, I might end up with some batches that are much bigger than other batches, right? That doesn't seem ideal. Every instance turns into multiple sequences. Can you make sure that we always run the same number of sequences, and then combine the results for each instance afterwards?

Reply:
In lines 114 I filter out these encodings into only the first sample for each example that contains an answer. I believe this should ensure that we have the same batch size every time.

Reply:
The batch sizes can still be variable if the answer isn't found in any of the samples for one or more example. I suggested an assert that prevents empty batches, but batches can still vary from full size to less than full size, right?

Reply:
I am more worried about the opposite case. You want to run a big evaluation (or an evaluation on a small GPU), and so you need to maximize the batch size. But one out of 100 batches ends up 50% bigger than the others, so now you have to make the batch size smaller for the entire evaluation and run at half speed, just to make that one batch fit.

Reply:
Yep this is an issue but we will remove this logic of running multiple samples per example once we switch over to the discussed truncation method.

Reply:
Aha I see, I didn't realize we were discussing the case where we examine all the sampled windows and not just the one with the answer in it. To address @dirkgr's concern, could we try batching by total token count instead of by number of examples in the batch. This would be fussy to implement because you could have to wait for more than one batch to go thru before resolving which of the sampled windows for a given instance wins out. But if it's done right you'll get perfect batch sizes for every batch except the tail one.